codegenerator/cli/templates/static/codegen/src/db/DbFunctionsImplementation.js

```diff
@@ -572,6 +502,69 @@ module.exports.deleteRolledBackEntityHistory = (
     );
   `;
 
+const Utils = require("envio/src/Utils.bs.js");
+
+module.exports.pruneStaleEntityHistory = (
+  sql,
+  entityName,
+  safeChainIdAndBlockNumberArray,
+) => {
+  const tableName = makeHistoryTableName(entityName);
+  return sql`
+  WITH first_change AS (
+    SELECT
+      MIN(serial) AS first_change_serial
+    FROM
+      public.${sql(tableName)}
+    WHERE
+      ${Utils.$$Array.interleave(
+    safeChainIdAndBlockNumberArray.map(
+      ({ chainId, blockNumber }) =>
+        sql`(entity_history_chain_id = ${chainId} AND entity_history_block_number > ${blockNumber})`,
+    ),
+    sql` OR `,
+  )}
+  ),
+  items_in_reorg_threshold AS (
+    SELECT DISTINCT
+      ON (id) *
+    FROM
+      public.${sql(tableName)}
+    WHERE
+      serial >= (SELECT serial FROM first_change)
+    ORDER BY
+      id,
+      serial ASC -- Select the row with the lowest serial per id
+  ),
+  -- Select all the previous history items for each id in the reorg threshold
+  previous_items AS (
+    SELECT
+      prev.id,
+      prev.serial
+    FROM
+      public.${sql(tableName)} prev
+    INNER JOIN
+      items_in_reorg_threshold r
+    ON
+      r.id = prev.id
+      AND
+      r.previous_entity_history_chain_id = prev.entity_history_chain_id
+      AND
+      r.previous_entity_history_block_number = prev.entity_history_block_number
+      AND
+      r.previous_entity_history_log_index = prev.entity_history_log_index
+  )
+  DELETE FROM
+    public.${sql(tableName)} eh
+  WHERE
+    -- Delete all entity history of entities that are not in the reorg threshold
+    eh.id NOT IN (SELECT id FROM items_in_reorg_threshold)
+    -- Delete all rows where id matches a row in previous_items but has a lower serial
+    OR 
+    eh.serial < (SELECT serial FROM previous_items WHERE previous_items.id = eh.id);
```

Comment:
The amazing advantage with this, is we only need to store history items for changes in the current reorg threshold + 1 historical change for each of those entities. Everything else can be cleared. 

On new inserts, a defaulted zero value historical item is copied in from the current entity table so we can always keep the history table quite slim.

Reply:
The only disadvantage is that this now has to be done as part of the batch transaction so we don't accidentally remove needed items due to async conditions. The main reason this query was running heavy before though is that there had to be at least one copy of every entity in the history table so it was quite large to do these expensive queries. Now it should be small based on changes within the reorg threshold of a chain.